cclient: Implement get_balances

howjmay opened the pull request:
Implement iota api get_balances
json_serializer.h:
        json_get_balances_serialize_request()
        json_get_balances_deserialize_response()

iota_api.c:
        iota_api_get_balances()

Implement int_array related functions.
json_serializer.c:
        json_array_to_int_array()
        int_array_to_json_array()

type.h: int_array_new()
        int_array_allocate()
        int_array_free()

request/get_balances.h:
        get_balances_req_free()
        get_balances_req_add_address()
        get_balances_req_add_tip()

response/get_blanaces.h:
        get_balances_res_free()
        get_balances_res_balances_at()
        get_balances_res_milestone_at()



# Test Plan:
CI must pass

howjmay committed:
cclient: Implement get_balances

Implement iota api get_balances
json_serializer.h:
        json_get_balances_serialize_request()
        json_get_balances_deserialize_response()

iota_api.c:
        iota_api_get_balances()

Implement int_array related functions.
json_serializer.c:
        json_array_to_int_array()
        int_array_to_json_array()

type.h: int_array_new()
        int_array_allocate()
        int_array_free()

request/get_balances.h:
        get_balances_req_free()
        get_balances_req_add_address()
        get_balances_req_add_tip()

response/get_blanaces.h:
        get_balances_res_free()
        get_balances_res_balances_at()
        get_balances_res_milestone_at()

howjmay commented:
@jserv

tsvisabo commented on cclient/types/types.c:
```diff
@@ -31,3 +31,27 @@ void char_buffer_free(char_buffer_t* in) {
+    return RC_OK;
+  }
+  in->array = (int*)malloc(sizeof(int) * (n + 1));
+  if (in->array == NULL) return RC_CCLIENT_OOM;
```
please add brackets

tsvisabo commented on cclient/types/types.c:
```diff
@@ -31,3 +31,27 @@ void char_buffer_free(char_buffer_t* in) {
+
+void int_array_free(int_array* in) {
+  if (in) {
+    if (in->array) free(in->array);
```
brackets :)

tsvisabo commented on cclient/serialization/tests/test_json.c:
```diff
@@ -419,11 +419,52 @@ void test_deserialize_get_inclusion_states(void) {
+      "[\"114544444\"], "
+      "\"references\": "
+      "[\"INRTUYSZCWBHGFGGXXPWRWBZACYAFGVRRP9VYEQJOHYD9URME"
+      "LKWAFYFMNTSP9MCHLXRGAFMBOZPZ9999\"], "
```
can you store these literals in  static variables or define them?

tsvisabo commented on cclient/serialization/json/json_serializer.c:
```diff
@@ -49,6 +49,44 @@ retcode_t utarray_to_json_array(UT_array* ut, cJSON* json_root,
+  if (cJSON_IsArray(json_item)) {
+    cJSON* current_obj = NULL;
+    int_array_allocate(in, cJSON_GetArraySize(json_item));
+    int* in_arr_p = in->array;
```
I think you better use `in->array` directly, because as much a as it seems not probable, if the underlying type is change from int to size_t then you cast to wrong type

oopsmonk commented on cclient/request/get_balances.c:
```diff
@@ -0,0 +1,34 @@
+    utarray_free(req->addresses);
+    utarray_free(req->tips);
+    free(req);
+    req = NULL;
```
no need to assign `NULL` to the copy of the pointer.

oopsmonk commented on cclient/serialization/json/json_serializer.c:
```diff
@@ -49,6 +49,44 @@ retcode_t utarray_to_json_array(UT_array* ut, cJSON* json_root,
+
+    cJSON_ArrayForEach(current_obj, json_item) {
+      if (current_obj->valuestring != NULL) {
+        *in_arr_p = atoi(current_obj->valuestring);
```
what if the value of string bigger than (2^31)-1 on a 32bit platform?

howjmay replied:
could I use the int* in int_array to point to a int array whose each element is each the char in the balances string?

howjmay committed:
Fix couple problems

types.c:
    add brackets

test_json.h:
    create this header file for defining macro for testing command text

json_serializer.c:
    remove int* in_arr_p and use in->array directly

request/get_balances.c:
    remove req = NULL

serialization/BUILD:
    add dependency for test_json.h

tsvisabo reviewed (approved):
Thanks @HowJMay  :)

tsvisabo committed:
Merge branch 'develop' into balance

howjmay committed:
implement type int_array_array for long int

To avoid program crashing if the return balance is a string
bigger than (2^31)-1 on a 32 bit platform, I implemented a new
type int_array_array for variable balances in
response/get_balances.h

The new type int_array_array store the balances in a int_array.
Each element in int_array.array is an element in balances
string.

Therefore, to meet the need above, I implemented
int_array_array related functions in cclient/type

response/get_balances:
    modified the var type of balances
    modified the return value type of get_balances_res_balances_at()

json_serializer.c:
    remove retcode_t json_array_to_int_array()
    implement retcode_t json_array_to_int_array_array()

json_serializer.h:
    modified func type of json_get_balances_serialize_request()
    modified func type of json_get_balances_deserialize_response()
cclient/type:
    implement int_array* int_array_array_new()
    implement retcode_t int_array_array_allocate()
    implement void int_array_array_free()
    imeplement char* int_array_to_string()
    implement int_array* string_to_int_array()

howjmay committed:
Merge branch 'balance_b' into balance
